openpype/lib/applications.py

```diff
@@ -1666,26 +1662,31 @@ def prepare_context_environments(data, env_group=None, modules_manager=None):
     system_settings = get_system_settings()
     data["project_settings"] = project_settings
     data["system_settings"] = system_settings
-    # Apply project specific environments on current env value
-    apply_project_environments_value(
-        project_name, data["env"], project_settings, env_group
-    )
 
     app = data["app"]
     context_env = {
         "AVALON_PROJECT": project_doc["name"],
-        "AVALON_ASSET": asset_doc["name"],
-        "AVALON_TASK": task_name,
         "AVALON_APP_NAME": app.full_name
     }
+    if asset_doc:
+        context_env["AVALON_ASSET"] = asset_doc["name"]
+        if task_name:
+            context_env["AVALON_TASK"] = task_name
 
     log.debug(
         "Context environments set:\n{}".format(
             json.dumps(context_env, indent=4)
         )
     )
     data["env"].update(context_env)
-    if not app.is_host:
+
+    # Apply project specific environments on current env value
+    # - apply them once the context environments are set
+    apply_project_environments_value(
+        project_name, data["env"], project_settings, env_group
+    )
+
+    if not app.is_host or not asset_doc or not task_name:
         return
```

Comment:
Is the idea that anything below this ONLY applies to any application that is a host? If so, it might be much clearer if we move that into a dedicated method `_prepare_host_context_environments` or alike so we can do:
```python
if app.is_host:
    self.prepare_host_context_environments(...)
```

Also, if asset_doc or task_name is not set but the application is a host we still intend to force those to be set, yes? If that's the case, then this should not just return but raise or log an error.
```python
if app.is_host:
    if not asset_doc or not task_name:
        raise RuntimeError("Missing required asset and task name for application launch with host support.")
    self.prepare_host_context_environments(...)
```

Reply:
I think you're overthinking this change. It just fill environment variables if the values are available all the time. It is not a reason to crash, and I don't think that asset and task is a "must have" requirement for a host to launch.

Reply:
> EDITED:
> The only change is, that if you would try to launch host without asset or task, you would at least know in which project it was launched.

Ok. In that case the testing notes should include a thorough check whether each DCC still works as expected for those cases. I'm pretty sure loaders, publishers, etc. will fail to work without asset + task available.

I'm pretty sure workfiles will fail for example. If that's the case I wouldn't expose it _without any warning or error_ as if things are all good just because in a far future "we might support it". To me it sounds like massive amounts of changes will be needed in the codebase to support a future where all DCCs can be launched without asset+task.. but I might be wrong, maybe it's just as easy as not setting `AVALON_ASSET` and `AVALON_TASK` and all will work. 🤞 

Also, as asked before - what's the expected behavior for Tools to be applied when launching an application without an asset/task since those are set per asset/folder? We'll need some fallback that in those cases then starts using the project defaults?